### DISCUSSION TITLE
Output validation

### DISCUSSION THREAD

**Discussion by matthiasbayer** [next-safe-action/next-safe-action#245] [Ideas]
Hey Edoardo, great project! Are there plans to add [output validation](https://trpc.io/docs/server/validators#output-validators)? 

I think it would be very handy and also make the migration from tRPC a bit easier:

```
// Current

const OutputSchema = z.object({ id: z.string(), foo: z.string(), bar: z.number() })
export const myAction = actionClient
  .schema(z.object({ id: z.string() }))
  .action(async ({ parsedInput: { id } }) => {
    return OutputSchema.parse(await Service.getEntity(id))
  })

// Suggested
export const myAction = actionClient
  .schema(z.object({ id: z.string() }))
  .schemaOutput(z.object({ id: z.string(), foo: z.string(), bar: z.number() }))
  .action(async ({ parsedInput: { id } }) => {
    return await Service.getEntity(id)
  })
```

Could add aliases (`input` => `schema` and `output` => `schemaOutput`) to make it a bit more intuitive.

Let me know what you think. Cheers!

**Comment by TheEdoRan** [comment]
> Are there plans to add [output validation](https://trpc.io/docs/server/validators#output-validators)?

Actually, yes. I've been wanting to implement this for a while now.

> Could add aliases (input => schema and output => schemaOutput) to make it a bit more intuitive.

I'm thinking of aliasing `inputSchema` to `schema` and add `outputSchema`. I really want to keep the "schema" word in the method name to avoid confusion about the input/output validation and parsing, which is performed when the action is called, and not in `inputSchema`/`outputSchema`.

**Comment by TheEdoRan** [comment]
Update: support will land in the library when #250 gets merged. Added `outputSchema` method to allow for optional action's output data validation. `schema` method is left unchanged to avoid confusion with weird aliases, should be good enough, at least for now.

**Comment by TheEdoRan** [comment]
Implemented in [v7.8.0](https://github.com/TheEdoRan/next-safe-action/releases/tag/v7.8.0)
